docs/design/interpreter/compiled-code-interop.md

```diff
@@ -0,0 +1,69 @@
+
+##### Interpreter entry
+
+Given we can't dynamically generate thunks that can be invoked, in order to generate an interpreter entry point we could reuse the functionality of fat pointers that is already used with native aot. Fat pointers can point to additional data, if a bit is set, rather than being actual function pointers. Instead of calling the pointer directly, the calling code will check for the most significant bit. If it is set, it will instead dereference the pointer and obtain the real function pointer together with the additional argument that is passed to the call. For the purpose of entering the interpreter, we would generate a fat pointer that has the target destination as a compiled IL wrapper for the signature in question together with the interpreter method pointer that is passed. The wrapper will obtain the pointer to the interpreter stack, will move all arguments there and call into the C++ interpreter path, passing the method and the address on the stack where the arguments have been written. When aot compiling an assembly, we would need to consider every single call as a potential entry to the interpreter and, if we deem it possible, we would generate an interpreter entry wrapper for the call signature in question.
+
+Native code has no knowledge of fat pointers, so we will need to explicitly generate small thunks for every single `UnmanagedCallersOnly` method. On mono, during app compilation, a build task scans the assembly for all `UnmanagedCallersOnly` methods and it dynamically generates a separate C method for each one of them. This method will have its own data that is later initialized with the function pointer for the cconv translation wrapper together with the interpreter method pointer argument. Rather than have this logic in special build tasks that dynamically generate C code, it might make more sense to  simply generate a special direct call wrapper for `UnmanagedCallersOnly` methods, that are not being aot compiled to the wasm image (this might represent a scenario that is just good to have, for interp-only, but not really mandatory since we can choose to always aot compile these methods).
+
+Given on wasm we might encounter situations where we don't have a necessary wrapper for a certain signature we would be in need for a fallback approach. While on browser we could rely on dynamic generation of wasm code, on wasi we have no alternative. For the rare cases where users would run into such scenarios, I think a simple approach would be for the runtime to report the missing signature when crashing and instruct the user to specify these signatures into a separate file, that can then be consumed by the wasm application build, so the additional wrappers are compiled.
```

Comment:
> While on browser we could rely on dynamic generation of wasm code, on wasi we have no alternative.

I would deprioritize any WASI-specific features. Again, great to have a way forward, but should not be prioritized at present.

Reply:
I know that jitting on wasm is dubious, but I feel the purpose of the interpreter on wasm would be more geared towards wasi, where there are dynamic code gen restrictions. On browser, in theory, RyuJIT could have a wasm backend and there would be no need for interpreter. So if we are to prototype something on wasm at some point, I think it would best to go with the solution that works everywhere, aka wasi. Anyhow I don't see wasm to be relevant for the prototype, just to have an ideea on how we can move forward there once we figure stuff out on arm64.